pkg/features/features.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package features
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/pingcap/tidb-operator/api/v2/core/v1alpha1"
+	meta "github.com/pingcap/tidb-operator/api/v2/meta/v1alpha1"
+)
+
+var defaultFeatureGates = NewFeatureGates()
+
+func Enabled(ns, name string, feat meta.Feature) bool {
+	return defaultFeatureGates.Enabled(ns, name, feat)
+}
+
+func Register(c *v1alpha1.Cluster) {
+	defaultFeatureGates.Register(c)
+}
+
+func Deregister(ns, name string) {
+	defaultFeatureGates.Deregister(ns, name)
+}
+
+type FeatureGates interface {
+	Enabled(ns, name string, feat meta.Feature) bool
+
+	Register(c *v1alpha1.Cluster)
+	Deregister(ns, name string)
+}
+
+type gates struct {
+	lock sync.Mutex
+
+	features map[string]featureSet
+}
```

Comment:
In scenarios with frequent reads and infrequent writes, sync.Map significantly reduces contention through lock-free read operations and partitioned write optimizations, demonstrating superior performance compared to a conventional map combined with sync.Mutex.

Reply:
I prefer `sync.RWMutex`. We need two `sync.Map` in this case and I feel it's complex and premature optimized. Maybe we can optimize it after we find any performance issues.